packages/server/src/createServer.ts

```diff
@@ -192,17 +198,23 @@ export function createServerWithConnection(connection: Connection) {
     }
   })
 
-  connection.onCompletion((docParams: TextDocumentPositionParams): CompletionItem[] => {
+  connection.onCompletion((docParams: CompletionParams): CompletionItem[] => {
+    // Make sure the client does not send use completion request for characters
+    // other than the dot which we asked for.
+    if (docParams.context?.triggerKind == CompletionTriggerKind.TriggerCharacter) {
+      if (docParams.context?.triggerCharacter != TRIGGER_CHARATER) {
+        return []
+      }
+    }
     let text = documents.get(docParams.textDocument.uri)?.getText()
     if (!text) {
       return []
     }
     logger.debug(text || '')
-    const candidates = complete(text, {
-  		line: docParams.position.line,
-  		column: docParams.position.character
-  	}, schema).candidates
-  	logger.debug(candidates.map(v => v.label).join(","))
+    let pos = { line: docParams.position.line, column: docParams.position.character }
+    var setting = SettingStore.getInstance().getSetting()
+    const candidates = complete(text, pos, schema, setting.jupyterLabMode).candidates
+    if (logger.isDebugEnabled()) logger.debug('onCompletion returns: ' + JSON.stringify(candidates))
```

Comment:
Why should this be in if condition?
As long as it uses `logger.debug`, it doesn't need to concern about log level on code lines.

Reply:
You are correct that `logger.debug` will get stubbed out when debugging is turned off. However computing the string passed to the `logger.debug` function will still happen. In cases where computing the string to be passed is costly you might want to place a condition before the `logger.debug`.

This is what I do in Java anyways. I assume the same is true in JavaScript.